optimade/server/entry_collections/elasticsearch.py

```diff
@@ -0,0 +1,210 @@
+        self.provider_fields = CONFIG.provider_fields.get(resource_mapper.ENDPOINT, [])
+
+        fields = list(self.get_attribute_fields())
+        fields.extend(resource_mapper.TOP_LEVEL_NON_ATTRIBUTES_FIELDS)
+        fields.extend(
+            [f"_{self.provider_prefix}_{field}" for field in self.provider_fields]
+        )
+        quantities = {}
+        for field in fields:
+            alias = resource_mapper.alias_for(field)
+            length_alias = resource_mapper.length_alias_for(field)
+
+            quantities[field] = Quantity(name=field, es_field=alias)
+            if length_alias is not None:
+                quantities[length_alias] = Quantity(name=length_alias)
+                quantities[field].length_quantity = quantities[length_alias]
+
+        if "elements" in quantities:
+            quantities["elements"].has_only_quantity = Quantity(name="elements_only")
+            quantities["elements"].nested_quantity = quantities["elements_ratios"]
+
+        if "elements_ratios" in quantities:
+            quantities["elements_ratios"].nested_quantity = quantities[
+                "elements_ratios"
+            ]
+
+        self.transformer = ElasticTransformer(quantities=quantities.values())
+
+        self.parser = LarkParser(
+            version=(0, 10, 1), variant="default"
+        )  # The ElasticTransformer only supports v0.10.1 as the latest grammar
+
+        self.index = index
+        body = index_definitions.get(index, None)
+        if body is None:
+            body = {
+                "mappings": {
+                    "doc": {
+                        "properties": {
+                            resource_mapper.alias_for(field): {"type": "keyword"}
+                            for field in fields
+                        }
+                    }
+                }
+            }
+
+        client.indices.create(index=self.index, ignore=400, body=body)
+
+    def __len__(self):
+        return Search(using=client, index=self.index).execute().hits.total
+
+    def __iter__(self):
+        raise NotImplementedError
+
+    def __contains__(self, entry):
+        raise NotImplementedError
+
+    def count(self, **kwargs):
+        raise NotImplementedError
+
+    def insert(self, data: List[EntryResource]) -> None:
+        def get_id(item):
+            if self.index == "links":
+                id = "%s-%s" % (item["id"], item["type"])
+            elif "id" in item:
+                id = item["id"]
+            elif "_id" in item:
+                # use the existing MongoDB ids in the test data
+                id = str(item["_id"])
+            else:
+                # ES will generate ids
+                id = None
+            item.pop("_id", None)
+            return id
+
+        bulk(
+            client,
+            [
+                dict(_index=self.index, _id=get_id(item), _type="doc", _source=item)
+                for item in data
+            ],
+        )
+
+    def find(
```

Comment:
I feel that much of the parameter handling (especially all the guard checks) could be done in the base class, as they are basically independent of the underlying db. I would suggest to implement `find(params)` in the base class and have an abstract method `_run_db_query(filter_query, page_limit, page_offset, sort_field, order)`.